File: services/modules/app-ports/src/main/scala/solutions/s4y/vocabla/app/ports/CreateEntryUseCase.scala
package solutions.s4y.vocabla.app.ports

import solutions.s4y.vocabla.app.ports
import solutions.s4y.vocabla.domain.identity.Identifier.given
import solutions.s4y.vocabla.domain.identity.{Identifier, IdentifierSchema}
import solutions.s4y.vocabla.domain.{Entry, Student, Tag}
import zio.schema.annotation.{caseName, description, fieldName, recordName}
import zio.schema.validation.Validation
import zio.schema.{DeriveSchema, Schema}
import zio.{Chunk, ZIO}

@description("Command to create a new vocabulary entry.")
final case class CreateEntryCommand(
    @description(
      "Entry to be added to the vocabulary."
    )
    entry: Entry,
    @description("IDs of tags to be associated with the entry.")
    tagIds: Chunk[Identifier[Tag]],
    @description("ID of the student who owns the entry.")
    ownerId: Identifier[Student]
)

object CreateEntryCommand:
  @description("Response containing the ID of the newly created entry.")
  final case class Response(
      @description("ID of the newly created entry.")
      entryId: Identifier[Entry]
  )
  given (using IdentifierSchema): Schema[CreateEntryCommand] =
    DeriveSchema.gen[CreateEntryCommand]

  given (using IdentifierSchema): Schema[Response] =
    DeriveSchema.gen[Response]

@description("Use case for creating a new vocabulary entry.")
trait CreateEntryUseCase:
  def apply[R](
      command: CreateEntryCommand
  ): ZIO[R, String, CreateEntryCommand.Response]

File: front_react/src/app-ports/entries/CreateEntryUseCase.ts
import type {Identifier} from '../../domain/identity/Identifier.ts';
import type {Tag} from '../../domain/Tag.ts';
import {Context, Effect} from 'effect';
import type {AppError} from '../errors/AppError.ts';
import type {Entry} from '../../domain/Entry.ts';

export type CreateEntryRequest = {
  readonly entry: Entry,
  readonly tagIds: Array<Identifier<Tag>>
}

export const CreateEntryRequest = (entry: Entry, tagIds: Array<Identifier<Tag>>): CreateEntryRequest => ({
  entry,
  tagIds
})

export type CreateEntryResponse = {
  readonly id: Identifier<Entry>;
}

export interface CreateEntryUseCase {
  createEntry(request: CreateEntryRequest): Effect.Effect<CreateEntryResponse, AppError>
}

export class CreateEntryUseCaseTag extends Context.Tag('CreateEntryUseCaseTag')<
  CreateEntryUseCaseTag,
  CreateEntryUseCase
>() {
}
